src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.BoolValueSet.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    using static BinaryOperatorKind;
+
+    internal static partial class ValueSetFactory
+    {
+        private class BoolValueSet : IValueSet<bool>
+        {
+            private readonly bool _hasFalse, _hasTrue;
+
+            internal static BoolValueSet AllValues = new BoolValueSet(hasFalse: true, hasTrue: true);
+            internal static BoolValueSet None = new BoolValueSet(hasFalse: false, hasTrue: false);
+            internal static BoolValueSet OnlyTrue = new BoolValueSet(hasFalse: false, hasTrue: true);
+            internal static BoolValueSet OnlyFalse = new BoolValueSet(hasFalse: true, hasTrue: false);
+
+            private BoolValueSet(bool hasFalse, bool hasTrue) => (_hasFalse, _hasTrue) = (hasFalse, hasTrue);
+
+            public static BoolValueSet Create(bool hasFalse, bool hasTrue)
+            {
+                switch (hasFalse, hasTrue)
+                {
+                    case (false, false):
+                        return None;
+                    case (false, true):
+                        return OnlyTrue;
+                    case (true, false):
+                        return OnlyFalse;
+                    case (true, true):
+                        return AllValues;
+                }
+            }
+
+            bool IValueSet.IsEmpty => !_hasFalse && !_hasTrue;
+
+            IValueSetFactory<bool> IValueSet<bool>.Factory => BoolValueSetFactory.Instance;
+
+            IValueSetFactory IValueSet.Factory => BoolValueSetFactory.Instance;
+
+            public bool Any(BinaryOperatorKind relation, bool value)
+            {
+                switch (relation, value)
+                {
+                    case (Equal, true):
+                        return _hasTrue;
+                    case (Equal, false):
+                        return _hasFalse;
+                    default:
+                        throw new ArgumentException("relation");
+                }
+            }
+
+            bool IValueSet.Any(BinaryOperatorKind relation, ConstantValue value) => value.IsBad || Any(relation, value.BooleanValue);
+
+            public bool All(BinaryOperatorKind relation, bool value)
+            {
+                switch (relation, value)
+                {
+                    case (Equal, true):
+                        return !_hasFalse;
```

Comment:
>return !_hasFalse; [](start = 24, length = 18)

The behavior on `BoolValueSet.None` is worth calling out, as it means that `All` is not a subset of `Any` as one might have expected.

Reply:
That is true for every kind of value set.  Such a comment belongs on the interface declarations for `All` and `Any`, and I'll make sure there is something there.

---
In reply to: [381019642](https://github.com/dotnet/roslyn/pull/41503#discussion_r381019642) [](ancestors = 381019642)